cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,219 @@
+package events
+
+import (
+	"io"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/xdr"
+)
+
+// Cursor represents the position of an event within the sequence of all
+// Soroban events. Soroban events are sorted by ledger sequence, transaction
+// index, operation index, and event index.
+type Cursor struct {
+	// Ledger is the sequence of the ledger which emitted the event.
+	Ledger uint32
+	// Tx is the index of the transaction within the ledger which emitted the event.
+	Tx uint32
+	// Op is the index of the operation within the transaction which emitted the event.
+	Op uint32
+	// Event is the index of the event within in the operation which emitted the event.
+	Event uint32
+}
+
+var (
+	// MinCursor is the smallest possible cursor
+	MinCursor = Cursor{}
+	// MaxCursor is the largest possible cursor
+	MaxCursor = Cursor{
+		Ledger: math.MaxUint32,
+		Tx:     math.MaxUint32,
+		Op:     math.MaxUint32,
+		Event:  math.MaxUint32,
+	}
+)
+
+type event struct {
+	contents xdr.ContractEvent
+	cursor   Cursor
+}
+
+func cmp(a, b uint32) int {
+	if a < b {
+		return -1
+	}
+	if a > b {
+		return 1
+	}
+	return 0
+}
+
+// Cmp compares two cursors.
+// 0 is returned if the c is equal to other.
+// 1 is returned if c is greater than other.
+// -1 is returned if c is less than other.
+func (c Cursor) Cmp(other Cursor) int {
+	if c.Ledger == other.Ledger {
+		if c.Tx == other.Tx {
+			return cmp(c.Event, other.Event)
+		}
+		return cmp(c.Tx, other.Tx)
+	}
+	return cmp(c.Ledger, other.Ledger)
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
+	lock            sync.RWMutex
+	events          []event
+	length          int
+	start           int
+	retentionWindow uint32
```

Comment:
Please add a comment or rename it to indicate the unit in which the retention window is expressed.